src/analysis/ppx_expand.ml

```diff
@@ -0,0 +1,134 @@
+    match expression.pexp_desc with
+    | Pexp_extension (loc, _) ->
+        atr := Some expression.pexp_loc;
+        check_at_pos loc.loc
+    | _ -> false
+  in
+  let expr (self : Ast_iterator.iterator) (expr : Parsetree.expression) =
+    match check_extension_node expr with
+    | true -> (
+        let expr (self : Ast_iterator.iterator) (exp : Parsetree.expression) =
+          match
+            expr.pexp_loc.loc_start.pos_cnum <= exp.pexp_loc.loc_start.pos_cnum
+            && exp.pexp_loc.loc_end.pos_cnum <= expr.pexp_loc.loc_end.pos_cnum
+          with
+          | true -> expression := exp :: !expression
+          | false -> Ast_iterator.default_iterator.expr self exp
+        in
+        let iterator = { Ast_iterator.default_iterator with expr } in
+        match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+    | false -> Ast_iterator.default_iterator.expr self expr
+  in
+  let signature_item (self : Ast_iterator.iterator)
+      (item_1 : Parsetree.signature_item) =
+    match check_signatures item_1.psig_desc with
+    | true -> (
+        let signature_item (self : Ast_iterator.iterator)
+            (item_2 : Parsetree.signature_item) =
+          match check_at_pos item_2.psig_loc && item_1 <> item_2 with
+          | true -> signature := item_2 :: !signature
+          | false -> Ast_iterator.default_iterator.signature_item self item_2
+        in
+        let iterator = { Ast_iterator.default_iterator with signature_item } in
+        match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+    | false -> Ast_iterator.default_iterator.signature_item self item_1
+  in
+  let structure_item (self : Ast_iterator.iterator)
+      (item_1 : Parsetree.structure_item) =
+    match check_structures item_1.pstr_desc with
+    | true -> (
+        let structure_item (self : Ast_iterator.iterator)
+            (item_2 : Parsetree.structure_item) =
+          match check_at_pos item_2.pstr_loc && item_1 <> item_2 with
+          | true -> structure := item_2 :: !structure
+          | false -> Ast_iterator.default_iterator.structure_item self item_2
+        in
+        let iterator = { Ast_iterator.default_iterator with structure_item } in
+        match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+    | false -> Ast_iterator.default_iterator.structure_item self item_1
+  in
+  let iterator =
+    { Ast_iterator.default_iterator with signature_item; structure_item; expr }
+  in
+  let _ =
+    match parsetree with
+    | `Interface si -> iterator.signature iterator si
+    | `Implementation str -> iterator.structure iterator str
+  in
+  match (!signature, !structure, !expression) with
```

Comment:
This matching makes me sad :-) 
Lets see if other refactoring might lead us to a nicer code organization...